macros/contexts/contextBoolean.pl

```diff
@@ -0,0 +1,253 @@
+## contextBoolean.pl
+
+sub _contextBoolean_init { context::Boolean::Init() }
+
+package context::Boolean;
+
+sub Init {
+	my $context = $main::context{Boolean} = Parser::Context->getCopy('Numeric');
+	$context->{name}            = 'Boolean';
+	$context->{parser}{Formula} = 'context::Boolean::Formula';
+	$context->{value}{Formula}  = 'context::Boolean::Formula';
+	$context->{parser}{Number}  = 'context::Boolean::Boolean';
+	$context->{value}{Boolean}  = 'context::Boolean::Value';
+	$context->{value}{Real}     = 'context::Boolean::Value';
+
+	$context->{pattern}{number}       = '^[01TtFf]$';
+	$context->{pattern}{signedNumber} = '^[01TtFf]$';
+	$context->flags->set(
+		NumberCheck => sub {
+			my $self = shift;
+			$self->{value}  = ($self->{value_string} =~ /[t1]/i) ? 1 : 0;
+			$self->{isOne}  = ($self->{value} == 1);
+			$self->{isZero} = ($self->{value} == 0);
+		}
+	);
+	$context->update;
+
+	my $orPrec  = $context->operators->get('+')->{precedence};
+	my $andPrec = $context->operators->get('*')->{precedence};
+	my $notPrec = $context->operators->get('u-')->{precedence};
+
+	## Disable unnecessary context stuff
+	$context->functions->clear();
+	$context->strings->clear();
+	$context->lists->clear();
+
+	## Define constants for 'True' and 'False'
+	$context->constants->are(
+		'T' =>
+			{ value => context::Boolean::Value->new(1), isConstant => 0, string => 'T', TeX => '\top', perl => 1 },
+		'F' =>
+			{ value => context::Boolean::Value->new(0), isConstant => 0, string => 'F', TeX => '\bot', perl => 0 },
+		# 0 and 1 are not recognized unless added as constants
+		# it might not be a bad idea anyhow?
+		'1'     => { alias => 'T' },
+		'0'     => { alias => 'F' },
+		'True'  => { alias => 'T' },
+		'False' => { alias => 'F' },
+	);
+
+	## Define our logic operators
+	$context->operators->are(
+		'or' => {
+			class         => 'context::Boolean::BOP::or',
+			precedence    => $orPrec,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' or ',
+			TeX           => '\vee ',
+			perl          => '||',
+			alteratives   => ['\x{2228}'],
+		},
+		'and' => {
+			class         => 'context::Boolean::BOP::and',
+			precedence    => $andPrec,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' and ',
+			TeX           => '\wedge ',
+			perl          => '&&',
+			alteratives   => ['\x{2229}'],
+		},
+		'xor' => {
+			class         => 'context::Boolean::BOP::xor',
+			precedence    => $andPrec,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' xor ',
+			perl          => '!=',
+			TeX           => '\oplus ',
+			alteratives   => [ '\x{22BB}', '\x{2295}' ],
+		},
+		# unary negation is only allowed if the '-' token is recognized, so make it BOP::undef
+		'-' => {
+			class         => 'Parser::BOP::undefined',
+			precedence    => $notPrec - 1,
+			associativity => 'left',
+			type          => 'both',
+			string        => '-',
+		},
```

Comment:
Okay, I get this. There might be an unintended consequence though.

Since the last PR, I've added some reduction rules -- and one relies on Parser::Item::isNeg (which is hardcoded to look for 'u-') to reduce a formula such as '!!p'. I suspect this change might break that.

Reply:
I found that issue because I had 'not' as the base-token for 'u-' (and '!', etc), so I switched 'u-' to be the token that holds the UOP class and everything else aliased to it instead.

Reply:
Rather than make an unnecessary `u-` operator, it is probably easier to add

``` perl
sub isNeg {return 1}
```

to the `context::Boolean::not` package, since that is the only unary operator, and it is the "negation" operator.

The `Parser::Item::isNeg` method probably needs to be reworked.  It should just return 0, and then in `Parser::UOP` is should be override with a version that checks the `uop` and for infinite `op`.  It might be valuable to get the name of the operator (or even operators) that count as negation from a context flag so that it is more generic.

There are still a number of places in the Parser/Value packages where some hard-coded things should be made more general.

Reply:
I was also thinking about reduction rules, and `-(-p) => p` is one, as would be `-(p and q) => -p or -q` and `-(p or q) => -p and -q`.  I'm not sure what else would be useful.

Reply:
I'll add this (revising `Parser::Item::isNeg`) to the list of action items (along with customizing context types)